tensorflow/java/src/gen/cc/java_defs.h

```diff
@@ -0,0 +1,319 @@
+  JavaDoc doc_;
+
+  explicit JavaMethod(const string& name) : name_(name) {}
+  JavaMethod(const string& name, const JavaType& type)
+    : name_(name), type_(type) {}
+
+  friend class Java;
+};
+
+/// \brief A factory of common Java definitions and other utilities.
+class Java {
+ public:
+  /// Returns the definition of a Java primitive type
+  static JavaType Type(const string& name) {
+    return JavaType(JavaType::PRIMITIVE, name);
+  }
+  /// Returns the definition of a Java class
+  static JavaType Class(const string& name, const string& package = "") {
+    return JavaType(JavaType::CLASS, name, package);
+  }
+  /// Returns the definition of a Java interface
+  static JavaType Interface(const string& name, const string& package = "") {
+    return JavaType(JavaType::INTERFACE, name, package);
+  }
+  /// Returns the definition of Java generic type parameter
+  static JavaType Generic(const string& name) {
+    return JavaType(JavaType::GENERIC, name);
+  }
+  /// Returns the definition of a Java wildcard type parameter (<?>)
+  static JavaType Wildcard() {
+    return JavaType(JavaType::GENERIC);
+  }
+  /// Returns the definition of a Java annotation
+  static JavaAnnot Annot(const string& type_name, const string& pkg = "") {
+    return JavaAnnot(JavaType(JavaType::ANNOTATION, type_name, pkg));
+  }
+  /// Returns the definition of Java variable
+  static JavaVar Var(const string& name, const JavaType& type) {
+    return JavaVar(name, type);
+  }
+  /// Returns the definition of periodic Java variable
+  static JavaVar PeriodicVar(const string& name, const JavaType& type) {
+    return JavaVar(name, type, true);
+  }
+  /// Returns the definition of Java class method
+  static JavaMethod Method(const string& name, const JavaType& return_type) {
+    return JavaMethod(name, return_type);
+  }
+  /// Returns the definition of a Java class constructor
+  static JavaMethod ConstructorFor(const JavaType& clazz) {
+    return JavaMethod(clazz.name());
+  }
+  /// Returns the definition of the class of "type" (Class<type>)
+  static JavaType ClassOf(const JavaType& type) {
+    return Class("Class").param(type);
+  }
+  /// Returns the definition of a list of "type" (List<type>)
+  static JavaType ListOf(const JavaType& type) {
+    return Interface("List", "java.util").param(type);
+  }
+  /// Returns the definition of a iteration of "type" (Iterable<type>)
+  static JavaType IterableOf(const JavaType& type) {
+    return Interface("Iterable").param(type);
+  }
+  /// Returns true if "type" is a wildcard type parameter (<?>)
+  static bool IsWildcard(const JavaType& type) {
+    return type.kind() == JavaType::GENERIC && type.name().empty();
+  }
+  /// Returns true if "type" is of a known collection type (only a few for now)
+  static bool IsCollection(const JavaType& type) {
+    return type.name() == "List" || type.name() == "Iterable";
+  }
+};
+
+// Templates implementation
+
+template <class TypeScanner>
```

Comment:
In fact, the ```GenericTypeScanner``` should be defined in the ```java_defs.h``` file and can be moved as a private utility in ```java_writer.cc```. There are others implementations (or there will) but are declared by simply using lambdas instead of a named type.